discovery/consul/consul.go

```diff
@@ -226,164 +290,225 @@ func (d *Discovery) Run(ctx context.Context, ch chan<- []*targetgroup.Group) {
 		default:
 		}
 
+		// Get the local datacenter first, if necessary.
+		err := d.getDatacenter()
 		if err != nil {
-			level.Error(d.logger).Log("msg", "Error refreshing service list", "err", err)
-			rpcFailuresCount.Inc()
 			time.Sleep(retryInterval)
 			continue
 		}
-		// If the index equals the previous one, the watch timed out with no update.
-		if meta.LastIndex == lastIndex {
-			continue
-		}
-		lastIndex = meta.LastIndex
-
-		// If the datacenter was not set from clientConf, let's get it from the local Consul agent
-		// (Consul default is to use local node's datacenter if one isn't given for a query).
-		if d.clientDatacenter == "" {
-			info, err := d.client.Agent().Self()
-			if err != nil {
-				level.Error(d.logger).Log("msg", "Error retrieving datacenter name", "err", err)
-				time.Sleep(retryInterval)
-				continue
-			}
-			d.clientDatacenter = info["Config"]["Datacenter"].(string)
-		}
-
-		// Check for new services.
-		for name := range srvs {
-			if !d.shouldWatch(name) {
-				continue
-			}
-			if _, ok := services[name]; ok {
-				continue // We are already watching the service.
-			}
+		// We are good to go.
+		return
+	}
+}
 
-			srv := &consulService{
-				client: d.client,
-				name:   name,
-				labels: model.LabelSet{
-					serviceLabel:    model.LabelValue(name),
-					datacenterLabel: model.LabelValue(d.clientDatacenter),
-				},
-				tagSeparator: d.tagSeparator,
-				logger:       d.logger,
+// Run implements the Discoverer interface.
+func (d *Discovery) Run(ctx context.Context, ch chan<- []*targetgroup.Group) {
+	d.initialize(ctx)
+
+	if len(d.watchedServices) == 0 || d.watchedTag != "" {
+		// We need to watch the catalog.
+		ticker := time.NewTicker(d.refreshInterval)
+		go func() {
+			// Watched services and their cancellation functions.
+			services := make(map[string]func())
+			var lastIndex uint64
+
+			for ; true; <-ticker.C {
```

Comment:
`for range ticker.C` is the usual way to spell this

Reply:
`for ; true; <-ticker.C`  is similar, but execute the loop once immediately to speed up initial target discover